### PR TITLE
Set go version to 1.17 to match CI

### LINKED ISSUE #10580: Build fails with go1.16 but works with 1.17, Go.mod points to go1.16
The commit 3841af9 updates helm to use go 1.17 on the CI but we have this on master:

https://github.com/helm/helm/blob/86a94f2f53b0519a3b22648a11f332c85cd8da1a/go.mod#L3

When building I encountered this error:

```
/data/go/pkg/mod/sigs.k8s.io/json@v0.0.0-20211020170558-c049b76a60c6/internal/golang/encoding/json/encode.go:1255:18: sf.IsExported undefined (type reflect.StructField has no field or method IsExported)
```

And found [here](https://github.com/kubernetes-sigs/controller-tools/issues/643#issuecomment-1004459362)  that updating to go1.17 fixes the issue.

I think, we need to update the go version in go.mod to 1.17.

### PR DESCRIPTION
**Pull request by smaftoul**
Closes #10580 

**What this PR does / why we need it**:

Bump go version to 1.17 to match CI. It's needed because building with 1.16 is broken and because the CI uses 1.17.

### CODE REVIEW

**Comment by smaftoul** [comment]
CI Has this error:
```
ERRO Running error: context loading failed: no go files to analyze
```

The PR doesn't contain any modification to go files, I guess this is normal (but probably should not fail ?)

**Comment by bacongobbler** [comment]
likely fixed once https://github.com/helm/helm/pull/10575 is merged - golint is deprecated.

**Comment by bacongobbler** [comment]
For reference, the version number listed in go.mod dictates the minimum version of Go you can use to compile the project. It does not have to match the same version in CI - in fact it can be several versions behind. However in the case described in #10580, it sounds like Helm cannot be compiled against Go 1.16 due to an issue with a transitive dependency, so bumping go.mod to a version that does compile seems fine to me.

**Review by bacongobbler — APPROVED** [review]
LGTM once we merge #10575 and this PR is rebased to ensure everything still works

**Comment by bacongobbler** [comment]
> I think, the CI should use the minimum version of Go you can use to compile the project, so this situation doesn't happen.

I agree that we should run the test suite against all supported versions of Go (and we'd appreciate a PR to do so next time!). But I don't think we should release new versions of Helm compiled against the minimum version. Newer versions of Go contain performance improvements and bug fixes, many of which will benefit Helm users, which is why we compile Helm against the latest stable version of Go.

**Comment by bacongobbler** [comment]
Can you please rebase? Thanks

**Comment by smaftoul** [comment]
@bacongobbler Done !

**Review by bacongobbler — APPROVED** [review]

**Pull request merged**
